Css contain improvements (#22494)
mdn/content: files/en-us/web/css/contain/index.md

{{CSSRef}}

The **`contain`** [CSS](/en-US/docs/Web/CSS) property indicates that an element and its contents are, as much as possible, independent from the rest of the document tree.
Containment allows the browser to calculate layout, style, paint, size, or any combination of these for a specific area of the DOM.

{{EmbedInteractiveExample("pages/css/contain.html")}}

> Changes within an element with containment applied are not propagated outside of the contained element to the rest of the page.
> The main benefit of this is that the browser does not have to re-render the DOM or page layout as often, leading to performance benefits.

Reviewer: ```suggestion
There are four types of CSS containment: size, layout,
style, and paint, which are set on the container. The property is a space-separated list of a subset of the five standard values and / or one of the two shorthand values. Changes to the contained properties within the container are not propagated outside of the contained element to the rest of the page.
The main benefit of containment is that the browser does not have to re-render the DOM or page layout as often, leading to small performance benefits during the rendering of static pages and greater performance benefits the more dynamic the application.
```

{{CSSRef}}

The **`contain`** [CSS](/en-US/docs/Web/CSS) property indicates that an element and its contents are, as much as possible, independent from the rest of the document tree.
Containment enables isolating a subsection of the DOM, providing performance benefits by limiting calculations of layout, style, paint, size, or any combination to a DOM subtree rather than the entire page. Containment can also be used to scope CSS counters and quotes.

{{EmbedInteractiveExample("pages/css/contain.html")}}

There are four types of CSS containment: size, layout, style, and paint, which are set on the container.

> The property is a space-separated list of a subset of the five standard values and / or one of the two shorthand values.

Reviewer: ```suggestion
The property is a space-separated list of a subset of the five standard values or one of the two shorthand values.
```

contain: unset;
```

### Values

The `contain` property can have any of the following values:

> - One of the keywords `none`, `strict`, or `content` **or**
> - One or more of the keywords `size` (or `inline-size`), `layout`, `style`, and `paint` in any order.

Reviewer: ```suggestion
- The keyword `none` **or**
- One or more of the space-separated keywords `size` (or `inline-size`), `layout`, `style`, and `paint` in any order **or**
 - One of the shorthand values `strict` or `content`
```

Author: Better, tnx

The keywords have the following meanings:

- `none`
  - : The element renders as normal, with no containment applied.
- `strict`
  - : All containment rules are applied to the element. This is equivalent to `contain: size layout paint style`.
- `content`
  - : All containment rules except `size` are applied to the element. This is equivalent to `contain: layout paint style`.
- `size`

>   - : Size containment is applied to the element. The size of the element can be computed in isolation, ignoring the child elements. This value cannot be combined with `inline-size`.

Reviewer: ```suggestion
  - : Size containment is applied to the element in both the inline and block directions. The size of the element can be computed in isolation, ignoring the child elements. This value cannot be combined with `inline-size`.
```

- `inline-size`
  - : Inline size containment is applied to the element. The inline size of the element can be computed in isolation, ignoring the child elements. This value cannot be combined with `size`.
- `layout`

>   - : The internal layout of the element is isolated from the rest of the page. That means nothing outside the element affects its internal layout, and vice versa.

Reviewer: ```suggestion
  - : The internal layout of the element is isolated from the rest of the page. This means nothing outside the element affects its internal layout, and vice versa.
```

- `style`

>   - : For properties that can have effects on more than just an element and its descendants, those effects don't escape the containing element. Counters and quotes are scoped to the element and its contents.

Reviewer: ```suggestion
  - : For properties that can affect more than just an element and its descendants, the effects don't escape the containing element. Counters and quotes are scoped to the element and its contents.
```

Consider the example below which shows how elements behave with and without layout containment applied:

```html
<div class="card" style="contain: layout;">
  <h2>Card 1</h2>
  <div class="fixed"><p>Fixed box 1</p></div>
  <div class="float"><p>Float box 1</p></div>
</div>
<div class="card">
  <h2>Card 2</h2>
  <div class="fixed"><p>Fixed box 2</p></div>
  <div class="float"><p>Float box 2</p></div>
</div>
<div class="card">
  <h2>Card 3</h2>
  <!-- ... -->
</div>
```

```css hidden
p {
  margin: 4px;
  padding: 4px;
}

h2 {
  margin-bottom: 4px;
  padding: 10px;
}

div {
  border-radius: 4px;
  box-shadow: 0 2px 4px 0 gray;
  padding: 6px;
  margin: 6px;
}
```

```css
.card {
  width: 70%;
  height: 90px;
}

.fixed {
  position: fixed;
  right: 10px;
  top: 10px;
  background: coral;
}

.float {
  float: left;
  margin: 10px;
  background: aquamarine;
}
```

The first card has layout containment applied, and its layout is isolated from the rest of the page.
We can reuse this card in other places on the page without worrying about layout recalculation of the other elements.
If floats overlap the card bounds, elements on the rest of the page are not affected.

> The browser only needs to recalculate the containing element's subtree, and not anything outside it.

Reviewer: ```suggestion
When the browser recalculates the containing element's subtree, only that element is recalculated. Nothing outside of the contained element needs to be recalculated.
```

Additionally, the fixed box uses the card as a layout container to position itself.

> The second card has has no containment applied.
> The fixed box in this card instead uses the root element as layout context and so the box has a fixed position in the top right corner of the page.
> Another side effect is that the layout in the third card looks broken because of the float that overlaps the second card's bounds.
> A portion of the DOM tree needs to be recalculated, and this interferes with the rest of the page layout.

Reviewer: ```suggestion
The second card has no containment applied.
The layout context for the fixed box in this card is the root element with the box having a fixed position in the top right corner of the page.
Without containment, the layout in the third card looks broken because the float overlaps the second card's bounds.
Without containment, when recalculation occurs, it is not limited to the container. This impacts performance and interferes with the rest of the page layout.
```

  <h2>Card 3</h2>
  <!-- ... -->
</div>
```

```css hidden
p {
  margin: 4px;
  padding: 4px;
}

h2 {
  margin-bottom: 4px;
  padding: 10px;
}

div {
  border-radius: 4px;
  box-shadow: 0 2px 4px 0 gray;
  padding: 6px;
  margin: 6px;
}
```

```css
.card {
  width: 70%;
  height: 90px;
}

.fixed {
  position: fixed;
  right: 10px;
  top: 10px;
  background: coral;
}

.float {
  float: left;
  margin: 10px;
  background: aquamarine;
}
```

The first card has layout containment applied, and its layout is isolated from the rest of the page.
We can reuse this card in other places on the page without worrying about layout recalculation of the other elements.
If floats overlap the card bounds, elements on the rest of the page are not affected.
When the browser recalculates the containing element's subtree, only that element is recalculated. Nothing outside of the contained element needs to be recalculated.
Additionally, the fixed box uses the card as a layout container to position itself.

The second card has no containment applied.
The layout context for the fixed box in this card is the root element with the box having a fixed position in the top right corner of the page.

> Without containment, the layout in the third card looks broken because the float overlaps the second card's bounds.

Reviewer: ```suggestion
Without containment, the floats overlap the second and third card's bounds, which is likely not the desired result.
```
This didn't look "broken" to me. is this what you are trying to say?

Author: The third card has some unexpected layout shifts - the float interferes with the H2 in this card. I'll reword to be more specific.

Author: Clearer now in bf3ad1988367c7a324c01886d5d884c86a099737

}
```

> When containment is applied in this example, the first closing quote ignores the inner span and uses the outer span's closing quote instead:

Reviewer: ```suggestion
Because of containment, the first closing quote ignores the inner span and uses the outer span's closing quote instead:
```
